tests/test_rate_limiter.py

```
@@ -0,0 +1,293 @@
+
+        await limiter.shutdown()
+
+    @pytest.mark.asyncio
+    async def test_execute_with_retry_exhausted(self, limiter):
+        """Test retry exhaustion."""
+        await limiter.initialize()
+
+        error = Exception("429 rate limit")
+        async_func = AsyncMock(side_effect=error)
+
+        with pytest.raises(Exception):
```

Ruff (0.15.2)

[warning] Do not assert blind exception: `Exception`

(B017)